### PR TITLE
Complete migration to Dialect/Feature flags + SQLServer support for Datetime

### PR DESCRIPTION
**Pull request by AdRiley**
### Pull Request Description

- Moves the remaining flags from the test code onto the Dialect flags.
- Adds support for SQLServer datetime functions

(Sorry this should have been two MRs)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [X] The documentation has been updated, if necessary.
- [X] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [X] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [X] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

### CODE REVIEW

File: test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso [hunk 1]
```diff
@@ -1270,12 +1269,15 @@ add_column_operation_specs suite_builder setup =
...
+                expected_new_size = case setup.flagged ..Char_Max_Size_After_Substring_Kept of
+                    False   -> Nothing
+                    True    -> 286
 
-                Test.with_clue "(Expected behaviour for substring size = "+setup.test_selection.char_max_size_after_substring.to_text+".) " <|
+                expected_behavior = case setup.flagged ..Char_Max_Size_After_Substring_Kept of
+                    False   -> "Reset"
+                    True    -> "Kept"
+
+                Test.with_clue "(Expected behaviour for substring size = "+expected_behavior+".) " <|
```
Comment (radeusgd) [inline test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso:1292]: What about the `Reduced` scenario?

**Review by radeusgd — COMMENTED** [review]
I'm slightly skeptical about in-memory having `flagged _ = True`. I know this was there earlier, when it was just simple basic features it seemed less problematic.

Now, since the `flagged` option checks various things, it feels a bit too restrictive. What if we add a feature that is only available in DBs but not in memory (rare but possible)? We'd probably make its flag negated for it to work with the `flagged _ = True`, but that is not good as negated flags often are less readable. Maybe we should make feature flags part of `Table` and allow in-memory backend to specify them in the same way as any other dialect?

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Dialect.enso [hunk 2]
```diff
@@ -685,131 +706,69 @@ replace args metadata =
 ## PRIVATE
 make_extract_as_int enso_name sql_name=enso_name =
     Base_Generator.lift_unary_op enso_name arg->
-        as_int32 <| SQL_Builder.code "EXTRACT(" ++ sql_name ++ " FROM " ++ arg ++ ")"
+        SQL_Builder.code sql_name ++ "(" ++ arg ++ ")"
```
Comment (radeusgd) [inline distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Dialect.enso]: After these changes this function is no longer `extract_as_int`, in fact if I'm not mistaken it is almost equivalent to `Base_Generator.make_function`, except that here the `lift_unary_op` will also check the number of arguments, whereas `make_function` does not do such a check. Still I'm not sure if this check is really needed, so IMHO we should replace it with `make_function`.

**Review by radeusgd — APPROVED** [review]
Looks great.

I feel like since we are now computing `date_diff` between months manually, we maybe should add more test cases for it - currently if I'm looking correctly we have **only 2** test cases for `date_diff` with `Month`.

This used to be reasonable - in most backends we were just direclty delegating to the DB's `datediff` method - so there's little point in testing the external vendor's database - we can assume that each vendor is doing extensive testing themselves, and even if we find an irregularity there isn't that much that we can do. This is also consistent with testing in `Base` - we have **many** more tests for `add_working_days` which is our custom implementation so we need to ensure its correctness vs e.g. `start_of` or `next` which just delegates to Java's date functions and there's little point to testing the Java implementations inside of our Enso tests.

However, since now we have a custom implementation for the month difference, it feels like it would be good to test a few more edge cases: like 29th of february, date_diff (1st Jan, 31 Dec) etc. To avoid making the tests run too long, these additional tests can run behind the `run_advanced_edge_case_tests`.

File: test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso (hunk 1 above)
Comment (AdRiley) [reply test/Table_Tests/src/Common_Table_Operations/Column_Operations_Spec.enso:1292]: I don't believe any engines used the Reduced scenario so I removed it.

**Comment by AdRiley** [comment]
> I'm slightly skeptical about in-memory having `flagged _ = True`. I know this was there earlier, when it was just simple basic features it seemed less problematic.
> 
> Now, since the `flagged` option checks various things, it feels a bit too restrictive. What if we add a feature that is only available in DBs but not in memory (rare but possible)? We'd probably make its flag negated for it to work with the `flagged _ = True`, but that is not good as negated flags often are less readable. Maybe we should make feature flags part of `Table` and allow in-memory backend to specify them in the same way as any other dialect?

I think there are 2 points here. 
1. The flags should be explicit. I very much agree with this and will make this change.
2. Where should the in-memory flags live? I don't think they are part of table. I could imagine creating an InMemory dialect module that just contains these flags. What do you think about that idea?

**Comment by AdRiley** [comment]
> Looks great.
> 
> I feel like since we are now computing `date_diff` between months manually, we maybe should add more test cases for it - currently if I'm looking correctly we have **only 2** test cases for `date_diff` with `Month`.
> 
> This used to be reasonable - in most backends we were just direclty delegating to the DB's `datediff` method - so there's little point in testing the external vendor's database - we can assume that each vendor is doing extensive testing themselves, and even if we find an irregularity there isn't that much that we can do. This is also consistent with testing in `Base` - we have **many** more tests for `add_working_days` which is our custom implementation so we need to ensure its correctness vs e.g. `start_of` or `next` which just delegates to Java's date functions and there's little point to testing the Java implementations inside of our Enso tests.
> 
> However, since now we have a custom implementation for the month difference, it feels like it would be good to test a few more edge cases: like 29th of february, date_diff (1st Jan, 31 Dec) etc. To avoid making the tests run too long, these additional tests can run behind the `run_advanced_edge_case_tests`.

Agree. I will add more tests. I think another dialect has a custom implementation too.

File: distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Dialect.enso (hunk 2 above)
Comment (AdRiley) [reply distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Dialect.enso]: Will fix

**Comment by radeusgd** [comment]
> 2\. Where should the in-memory flags live? I don't think they are part of table. I could imagine creating an InMemory dialect module that just contains these flags. What do you think about that idea?

Hmm, I think you are right that they don't necessarily need to live in `Standard.Table`. I think they can indeed continue to exist in `Table_Tests` - as they are only needed for testing. Separating them out into a separate file though will probably make things clearer.

**Comment by AdRiley** [comment]
> > 2. Where should the in-memory flags live? I don't think they are part of table. I could imagine creating an InMemory dialect module that just contains these flags. What do you think about that idea?
> 
> Hmm, I think you are right that they don't necessarily need to live in `Standard.Table`. I think they can indeed continue to exist in `Table_Tests` - as they are only needed for testing. Separating them out into a separate file though will probably make things clearer.

I'm going to do this as a follow up MR as I want to merge this one and raise the next one that builds on top of this.

File: distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso [hunk 3]
```diff
@@ -194,6 +197,26 @@ type Redshift_Dialect
...
+            Dialect_Flag.Order_By_Unicode_Normalization_By_Default -> True
+            Dialect_Flag.Allows_Mixed_Type_Comparisons -> False
+            Dialect_Flag.Supports_Unicode_Normalization -> False
+            Dialect_Flag.NaN_Non_Comparable -> True
+            Dialect_Flag.Distinct_Returns_First_Row_From_Group_If_Ordered -> True
+            Dialect_Flag.Date_Time -> True
+            Dialect_Flag.Text_Length_Limited_Columns -> False
+            Dialect_Flag.Fixed_Length_Text_Columns -> False
+            Dialect_Flag.Length_Restricted_Text_Columns -> True
+            Dialect_Flag.Removes_Trailing_Whitespace_Casting_From_Char_To_Varchar -> False
```
Comment (GregoryTravis) [inline distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso:211]: :(

File: distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso [hunk 4]
```diff
@@ -194,6 +197,26 @@ type Redshift_Dialect
...
             Dialect_Flag.Supports_Infinity -> True
             Dialect_Flag.Case_Sensitive_Text_Comparison -> True
+            Dialect_Flag.Supports_Sort_Digits_As_Numbers -> False
+            Dialect_Flag.Case_Insensitive_Ordering -> False
+            Dialect_Flag.Order_By_Unicode_Normalization_By_Default -> True
+            Dialect_Flag.Allows_Mixed_Type_Comparisons -> False
+            Dialect_Flag.Supports_Unicode_Normalization -> False
+            Dialect_Flag.NaN_Non_Comparable -> True
+            Dialect_Flag.Distinct_Returns_First_Row_From_Group_If_Ordered -> True
+            Dialect_Flag.Date_Time -> True
```
Comment (GregoryTravis) [inline distribution/lib/Standard/AWS/0.0.0-dev/src/Database/Redshift/Internal/Redshift_Dialect.enso:207]:
```suggestion
            Dialect_Flag.Supports_Date_Time -> True
```

**Review by GregoryTravis — APPROVED** [review]

**Pull request merged**